### PR TITLE
Plugins: add a default 'plugins' directory in the config dir

### PR DESCRIPTION
**Pull request by darosior**
Having a `plugins` directory in the configuration path can be very handy instead of manually adding `--plugin-dir=` or `--plugin=`, and it's easier to have it in the configuration directory than in the installation one.

This PR adds a function that will, if it exists, iterate through `<path_to_config_dir>/plugins` and add, without failing on error, each directory it contains as a new `plugin-dir`.

It permits to, for instance :
```shell
git clone https://github.com/lightningd/plugins ~/.lightning/plugins
```
and have (once requirements installed) all the community curated list of plugins available.

### CODE REVIEW

**Comment by cdecker** [comment]
> It permits to, for instance :
> 
> ```shell
> git clone https://github.com/lightningd/plugins ~/.lightning/plugins
> ```

That likely is not what you want to do anyway, since at least that repository contains quite a few highly experimental plugins (and some that will kill the daemon if not configured correctly, see `autoreload`). So selectively adding plugins might be good, but advocating for entire collections to be cloned into the plugin dir seems like a bad idea.

**Comment by cdecker** [comment]
Also you only have a single commit that sorts commands?

**Comment by darosior** [comment]
Ooops no that's a bad one

**Comment by darosior** [comment]
> > It permits to, for instance :
> > ```shell
> > git clone https://github.com/lightningd/plugins ~/.lightning/plugins
> > ```
> 
> That likely is not what you want to do anyway, since at least that repository contains quite a few highly experimental plugins (and some that will kill the daemon if not configured correctly, see `autoreload`). So selectively adding plugins might be good, but advocating for entire collections to be cloned into the plugin dir seems like a bad idea.

Yeah it is for now (as I experienced when testing ^^) but it was more to demonstrate that plugin setup would be easier.

*I messed up with my commit, rewriting it i'll push it in a second*

**Comment by darosior** [comment]
Ok I rewrote it but adding the `default_dir` as a parameter of the function (you always have better ideas when you wrote a code twice ^^)

**Comment by darosior** [comment]
Test which are failing on Travis pass on my side.

File: lightningd/plugin.c [hunk 1]
```diff
@@ -900,6 +900,21 @@ void clear_plugins(struct plugins *plugins)
...
+	if (d) {
+		struct dirent *di;
+		while ((di = readdir(d)) != NULL) {
+			if (streq(di->d_name, ".") || streq(di->d_name, ".."))
+				continue;
+			add_plugin_dir(plugins, path_join(NULL, default_dir, di->d_name), true);
+		}
+	}
+	closedir(d);
+    tal_free(default_dir);
```
Comment (rustyrussell) [inline lightningd/plugin.c]: Indent.  Also, it's an anti-pattern to take ownership of a parameter's lifetime like this without explicit indication to the caller.  I would just use tmpctx instead of NULL here, and in the caller.

Comment (darosior) [reply lightningd/plugin.c]: I added that `tal_free` in a hurry to make travis pass, I wasn't aware of : 
```
* allocate from.  We have a magic convenience context
* `tmpctx` for temporary allocations like this.
```
it's useful ^^.

**Comment by darosior** [comment]
@rustyrussell amended the changes, thanks for the review.

**Comment by m-schmoock** [comment]
Nice idea. Just tried.

Two issues:
- We need documentation about this.
- It does not work for me
     - I tried copying the plugins directories into the config dir, so that the path was like this: `~/.lightning/summary/summary.py`
     - I tried to symlink the directories into it

**Comment by darosior** [comment]
> Nice idea. Just tried.
> 
> Two issues:
> 
> * We need documentation about this. `doc/PLUGINS.md` and in the plugin repo.
> * It does not work for me

I'll add some.

>   * I tried copying the plugins directories into the config dir, so that the path was like this: `~/.lightning/summary/summary.py`
>   * I tried to symlink the directories into it
> 
> In both cases the daemon did not recognize the plugins

The path is `~/.lightning/plugins/<your_plugin_dir>`, it is so to separate plugins files (so it doesn't get messy if a plugin has multiple files).

**Comment by m-schmoock** [comment]
Ah, thanks for the clarificaion :D

**Comment by m-schmoock** [comment]
@darosior 
For the plugin repo I did this doc PR: https://github.com/lightningd/plugins/pull/32

**Comment by darosior** [comment]
> @darosior
> For the plugin repo I did this doc PR: [lightningd/plugins#32](https://github.com/lightningd/plugins/pull/32)

Great ! :D

**Comment by fiatjaf** [comment]
What if the plugin is just one executable? It should be possible to just put it in `~/.lightning/plugins/`, why not?

**Comment by darosior** [comment]
> What if the plugin is just one executable? It should be possible to just put it in `~/.lightning/plugins/`,

Why not just create a directory for it ?

> why not?

I think it can become messy with many plugins (especially if plugins comport multiple files).

**Comment by darwin** [comment]
Just FYI. My dockerized lightningd started segfaulting[1] on startup without changes on my side. Git bisect pinpointed it to a497295f92b9b15d718aba9998f8da3f140f55c7. 

I'm planning to investigate it further during this week.

[1] https://travis-ci.org/darwin/simverse/builds/535175428#L2948-L2955

**Comment by darwin** [comment]
I was unable to get better stack trace on core dump, so I just bisected the commit line by line.

The culprit is `closedir` call with NULL `d` ("plugins" default_dir does not exist in my case).
https://github.com/ElementsProject/lightning/pull/2647/files#diff-d7945e0ea0fc3a128a17dec5e0ff98ddR914

I guess it works for most environments because it depends on libc. In my docker container I have Alpine Linux and this crashes it.

I did a bit of googling and this discussion might be relevant:
https://lists.freebsd.org/pipermail/freebsd-standards/2014-January/002704.html

TL;DR; passing NULL into closedir is not safu

**Comment by darosior** [comment]
> TL;DR; passing NULL into closedir is not safu

My bad, thank you for the investigation.
